x/stakeibc/keeper/msg_server_create_trade_route.go

```diff
@@ -2,81 +2,187 @@ package keeper
+// Gov tx to to register a trade route that swaps reward tokens for a different denom
+//
+// Example proposal:
+//
+//		{
+//		   "title": "Create a new trade route for host chain X",
+//		   "metadata": "Create a new trade route for host chain X",
+//		   "summary": "Create a new trade route for host chain X",
+//		   "messages":[
+//		      {
+//		         "@type": "/stride.stakeibc.MsgCreateTradeRoute",
+//		         "authority": "stride10d07y265gmmuvt4z0w9aw880jnsr700jefnezl",
+//
+//		         "stride_to_host_connection_id": "connection-0",
+//				 "stride_to_reward_connection_id": "connection-1",
+//			     "stride_to_trade_connection_id": "connection-2",
+//
+//				 "host_to_reward_transfer_channel_id": "channel-0",
+//				 "reward_to_trade_transfer_channel_id": "channel-1",
+//			     "trade_to_host_transfer_channel_id": "channel-2",
+//
+//				 "reward_denom_on_host": "ibc/rewardTokenXXX",
+//				 "reward_denom_on_reward": "rewardToken",
+//				 "reward_denom_on_trade": "ibc/rewardTokenYYY",
+//				 "host_denom_on_trade": "ibc/hostTokenZZZ",
+//				 "host_denom_on_host": "hostToken",
+//
+//				 "pool_id": 1,
+//				 "min_swap_amount": 10000000,
+//				 "max_swap_amount": 1000000000
+//			  }
+//		   ],
+//		   "deposit": "2000000000ustrd"
+//	   }
+//
+// >>> strided tx gov submit-proposal {proposal_file.json} --from wallet
+func (ms msgServer) CreateTradeRoute(goCtx context.Context, msg *types.MsgCreateTradeRoute) (*types.MsgCreateTradeRouteResponse, error) {
 	ctx := sdk.UnwrapSDKContext(goCtx)
-	k.Logger(ctx).Info(fmt.Sprintf("create trade route: %s", msg.String()))
+	if ms.authority != msg.Authority {
+		return nil, errorsmod.Wrapf(govtypes.ErrInvalidSigner, "invalid authority; expected %s, got %s", ms.authority, msg.Authority)
+	}
+
+	// Validate trade route does not already exist for this denom
+	_, found := ms.Keeper.GetTradeRoute(ctx, msg.RewardDenomOnHost, msg.HostDenomOnHost)
+	if found {
+		return nil, errorsmod.Wrapf(types.ErrTradeRouteAlreadyExists,
+			"startDenom: %s, endDenom: %s", msg.RewardDenomOnHost, msg.HostDenomOnHost)
+	}
 
-	// get our addresses, make sure they're valid
-	_, err := sdk.AccAddressFromBech32(msg.Creator)
+	// Confirm the host chain exists and the withdrawal address has been initialized
+	hostZone, err := ms.Keeper.GetActiveHostZone(ctx, msg.HostChainId)
 	if err != nil {
-		return nil, errorsmod.Wrapf(sdkerrors.ErrInvalidAddress, "creator address is invalid: %s. err: %s", msg.Creator, err.Error())
+		return nil, err
+	}
+	if hostZone.WithdrawalIcaAddress == "" {
+		return nil, errorsmod.Wrapf(types.ErrICAAccountNotFound, "withdrawal account not initialized on host zone")
 	}
 
+	// Register the new ICA accounts
 	hostICA := types.ICAAccount{
 		ChainId:      msg.HostChainId,
 		Type:         types.ICAAccountType_WITHDRAWAL,
-		ConnectionId: msg.HostConnectionId,
-		Address:      msg.HostIcaAddress,
```

Comment:
Why don't we include the address here as hostZone.WithdrawalIcaAddress since we just checked that it is initialized already directly above?